File: demo-android/app/src/test/java/com/example/demo/pact/PactFixtures.kt
package com.example.demo.pact

import au.com.dius.pact.consumer.MockServer
import au.com.dius.pact.consumer.dsl.DslPart
import au.com.dius.pact.consumer.dsl.LambdaDsl
import au.com.dius.pact.consumer.dsl.PactDslJsonBody
import com.example.demo.data.model.ErrorResponse
import com.example.demo.data.model.Task
import com.example.demo.data.model.TaskPriority
import com.example.demo.data.model.TaskRequest
import com.example.demo.data.model.TaskStatus
import com.example.demo.data.remote.ApiClient
import com.example.demo.data.remote.TaskApi

object PactFixtures {
    const val CONSUMER = "demo-android"
    const val TASK_ID = "507f1f77bcf86cd799439011"
    const val TIMESTAMP_PATTERN = "^\\d{4}-\\d{2}-\\d{2}T.*$"

    fun createTaskApi(mockServer: MockServer): TaskApi =
        ApiClient.createTaskApi("${mockServer.getUrl()}/v1/")

    fun createTaskRequest(): TaskRequest = TaskRequest(
        title = "Prepare release notes",
        description = "Document release tasks",
        status = TaskStatus.TODO,
        priority = TaskPriority.MEDIUM
    )

    fun updateTaskRequest(): TaskRequest = TaskRequest(
        title = "Prepare release notes - updated",
        description = "Document release tasks in detail",
        status = TaskStatus.IN_PROGRESS,
        priority = TaskPriority.HIGH
    )

    fun createTaskRequestBody(): DslPart = PactDslJsonBody()
        .valueFromProviderState(TaskRequest::title.name, "taskTitle", createTaskRequest().title)
        .stringValue(TaskRequest::description.name, createTaskRequest().description!!)
        .stringValue(TaskRequest::status.name, createTaskRequest().status.name)
        .stringValue(TaskRequest::priority.name, createTaskRequest().priority.name)

    fun updateTaskRequestBody(): DslPart = PactDslJsonBody()
        .valueFromProviderState(TaskRequest::title.name, "updatedTitle", updateTaskRequest().title)
        .stringValue(TaskRequest::description.name, updateTaskRequest().description!!)
        .stringValue(TaskRequest::status.name, updateTaskRequest().status.name)
        .stringValue(TaskRequest::priority.name, updateTaskRequest().priority.name)

    fun taskResponseBody(titleExample: String = "Prepare release notes"): DslPart = PactDslJsonBody()
        .stringMatcher(Task::id.name, "^[a-f0-9]{24}$", TASK_ID)
        .stringType(Task::title.name, titleExample)
        .stringType(Task::description.name, "Document release tasks")
        .stringType(Task::status.name, TaskStatus.TODO.name)
        .stringType(Task::priority.name, TaskPriority.MEDIUM.name)
        .stringMatcher(Task::createdDate.name, TIMESTAMP_PATTERN, "2026-04-26T09:00:00.000Z")
        .stringMatcher(Task::updatedDate.name, TIMESTAMP_PATTERN, "2026-04-26T09:00:00.000Z")

    fun taskListResponseBody(): DslPart = LambdaDsl.newJsonArrayMinLike(1) { array ->
        array.`object` { item ->
            item.stringMatcher(Task::id.name, "^[a-f0-9]{24}$", TASK_ID)
            item.stringType(Task::title.name, "Prepare release notes")
            item.stringType(Task::description.name, "Document release tasks")
            item.stringType(Task::status.name, TaskStatus.TODO.name)
            item.stringType(Task::priority.name, TaskPriority.MEDIUM.name)
            item.stringMatcher(Task::createdDate.name, TIMESTAMP_PATTERN, "2026-04-26T09:00:00.000Z")
            item.stringMatcher(Task::updatedDate.name, TIMESTAMP_PATTERN, "2026-04-26T09:00:00.000Z")
        }
    }.build()

    fun duplicateTitleErrorBody(exampleMessage: String): DslPart = PactDslJsonBody()
        .stringMatcher(
            ErrorResponse::message.name,
            "^Task with title '.*' already exists$",
            exampleMessage
        )
}

File: demo-ios/DemoPactTests/PactFixtures.swift
import Foundation
import PactSwift
@testable import Demo

enum PactFixtures {
    static let consumer = "demo-ios"
    static let pactsDirectory: URL = {
        if let configured = ProcessInfo.processInfo.environment["PACT_OUTPUT_DIR"] {
            let url = URL(fileURLWithPath: configured, isDirectory: true)
            try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
            return url
        }
        let url = FileManager.default.temporaryDirectory.appendingPathComponent("pacts", isDirectory: true)
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }()

    static func mockService(provider: String) -> MockService {
        MockService(
            consumer: consumer,
            provider: provider,
            writePactTo: pactsDirectory
        )
    }
    static let taskID = "507f1f77bcf86cd799439011"
    static let timestampPattern = #"^\d{4}-\d{2}-\d{2}T.*$"#
    static let emptyProviderState = ProviderState(description: "", params: [:])

    static func makeTaskAPI(baseURL: String) -> TaskAPI {
        APIClient.makeTaskAPI(baseURL: apiBaseURL(from: baseURL))
    }

    static func apiBaseURL(from baseURL: String) -> URL {
        URL(string: "\(baseURL)/v1/")!
    }

    static let createTaskRequest = TaskRequest(
        title: "Prepare release notes",
        description: "Document release tasks",
        status: .todo,
        priority: .medium
    )

    static let updateTaskRequest = TaskRequest(
        title: "Prepare release notes - updated",
        description: "Document release tasks in detail",
        status: .inProgress,
        priority: .high
    )

    static func taskResponseBody(titleExample: String = "Prepare release notes") -> [String: Any] {
        [
            "id": Matcher.SomethingLike(taskID),
            "title": Matcher.SomethingLike(titleExample),
            "description": Matcher.SomethingLike("Document release tasks"),
            "status": Matcher.SomethingLike(TaskStatus.todo.rawValue),
            "priority": Matcher.SomethingLike(TaskPriority.medium.rawValue),
            "createdDate": Matcher.RegexLike(value: "2026-04-26T09:00:00.000Z", pattern: timestampPattern),
            "updatedDate": Matcher.RegexLike(value: "2026-04-26T09:00:00.000Z", pattern: timestampPattern),
        ]
    }

    static func createTaskRequestBody() -> [String: Any] {
        [
            "title": Matcher.SomethingLike(createTaskRequest.title),
            "description": Matcher.SomethingLike(createTaskRequest.description!),
            "status": Matcher.SomethingLike(createTaskRequest.status.rawValue),
            "priority": Matcher.SomethingLike(createTaskRequest.priority.rawValue),
        ]
    }

    static func updateTaskRequestBody() -> [String: Any] {
        [
            "title": Matcher.SomethingLike(updateTaskRequest.title),
            "description": Matcher.SomethingLike(updateTaskRequest.description!),
            "status": Matcher.SomethingLike(updateTaskRequest.status.rawValue),
            "priority": Matcher.SomethingLike(updateTaskRequest.priority.rawValue),
        ]
    }

    static func taskListResponseBody() -> Matcher.EachLike {
        Matcher.EachLike(taskResponseBody())
    }

    static func duplicateTitleErrorBody(exampleMessage: String) -> [String: Any] {
        [
            "message": Matcher.RegexLike(
                value: exampleMessage,
                pattern: "^Task with title '.*' already exists$"
            ),
        ]
    }
}
